File: algorithms/code/leetcode/lc125_valid_palindrome/lc125_valid_palindrome.java
package leetcode.lc125_valid_palindrome;

/**
 * https://leetcode.com/problems/valid-palindrome/ #easy
 */
public final class LC125ValidPalindrome {
    public boolean isPalindrome(final String s) {
        if (s == null || s.length() == 0) {
            return true;
        }
        int i = 0;
        int j = s.length() - 1;
        while (i <= j) {
            char a = s.charAt(i);
            if (!Character.isLetterOrDigit(a)) {
                i++;
                continue;
            }
            char b = s.charAt(j);
            if (!Character.isLetterOrDigit(b)) {
                j--;
                continue;
            }
            if (Character.toUpperCase(a) != Character.toUpperCase(b)) {
                return false;
            } else {
                i++;
                j--;
            }
        }
        return true;
    }
}
package leetcode.lc125_valid_palindrome;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

public final class LC125ValidPalindromeTests {
    @Test
    public void testAManAPlanACanalPanama() throws Exception {
        assertTrue(new LC125ValidPalindrome().isPalindrome("A man, a plan, a canal: Panama"));
    }

    @Test
    public void testRaceACar() throws Exception {
        assertFalse(new LC125ValidPalindrome().isPalindrome("race a car"));
    }

    @Test
    public void testAva() throws Exception {
        assertTrue(new LC125ValidPalindrome().isPalindrome("Ava"));
    }

    @Test
    public void testBurger() throws Exception {
        assertFalse(new LC125ValidPalindrome().isPalindrome("burger"));
    }

    @Test
    public void testNothing() throws Exception {
        assertTrue(new LC125ValidPalindrome().isPalindrome(null));
        assertTrue(new LC125ValidPalindrome().isPalindrome(""));
    }
}
```rust
fn is_palindrome(s: &str) -> bool {
    if s.is_empty() {
        return true;
    }

    let mut i = 0;
    let mut j = s.len() - 1;

    while i <= j {
        if !s.chars().nth(i).unwrap().is_ascii_graphic() {
            i += 1;
            continue;
        }
        if !s.chars().nth(j).unwrap().is_ascii_graphic() {
            j -= 1;
            continue;
        }
        if s.chars().nth(i).unwrap().to_ascii_uppercase() != s.chars().nth(j).unwrap().to_ascii_uppercase() {
            return false;
        } else {
            i += 1;
            j -= 1;
        }
    }

    true
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_a_man_a_plan_a_canal_panama() {
        assert!(is_palindrome("A man, a plan, a canal: Panama"));
    }

    #[test]
    fn test_race_a_car() {
        assert!(!is_palindrome("race a car"));
    }

    #[test]
    fn test_ava() {
        assert!(is_palindrome("Ava"));
    }

    #[test]
    fn test_burger() {
        assert!(!is_palindrome("burger"));
    }

    #[test]
    fn test_nothing() {
        assert!(is_palindrome(""));
        assert!(is_palindrome(null()));
    }
}
```

File: algorithms/code/leetcode/lc125_valid_palindrome/lc125_valid_palindrome.py
#!/usr/bin/env python3
# https://leetcode.com/problems/valid-palindrome/

import unittest


class Solution:
    def isPalindrome(self, s: str) -> bool:
        if len(s) == 0:
            return True
        i = 0
        j = len(s) - 1
        while i <= j:
            a = s[i]
            if not a.isalnum():
                i += 1
                continue
            b = s[j]
            if not b.isalnum():
                j -= 1
                continue
            if a.upper() != b.upper():
                return False
            i += 1
            j -= 1
        return True


class TestCode(unittest.TestCase):
    def test_a_man_a_plan_a_canal_Panama(self):
        self.assertTrue(
            Solution().isPalindrome('A man, a plan, a canal: Panama')
        )

    def test_race_a_car(self):
        self.assertFalse(Solution().isPalindrome('race a car'))

    def test_ava(self):
        self.assertTrue(Solution().isPalindrome('Ava'))

    def test_burger(self):
        self.assertFalse(Solution().isPalindrome('burger'))

    def test_nothing(self):
        self.assertTrue(Solution().isPalindrome(''))
